File: sdks/sdk-php/src/OpenPayClient.php
<?php

declare(strict_types=1);

namespace OpenPay;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\RequestException;
use OpenPay\Auth\HmacSigner;
use OpenPay\Exceptions\ApiException;
use OpenPay\Exceptions\AuthenticationException;
use OpenPay\Exceptions\OpenPayException;
use OpenPay\Resources\Payments;
use OpenPay\Resources\Webhooks;

/**
 * Open Pay API Client.
 *
 * Usage:
 *   $client = new OpenPayClient('ak_live_xxx.sk_live_yyy', [
 *       'base_url' => 'https://olp-api.nipuntheekshana.com',
 *   ]);
 *
 *   $payment = $client->payments->create([
 *       'amount' => '1000.00',
 *       'currency' => 'LKR',
 *       'merchantTradeNo' => 'ORDER-123',
 *   ]);
 */
class OpenPayClient
{
    private string $keyId;
    private string $secret;
    private string $baseUrl;
    private Client $http;

    public readonly Payments $payments;
    public readonly Webhooks $webhooks;

    /**
     * @param string $apiKey Compound API key: "ak_{env}_{id}.sk_{env}_{secret}"
     * @param array{base_url?: string, timeout?: float} $options
     */
    public function __construct(string $apiKey, array $options = [])
    {
        $parsed = HmacSigner::parseApiKey($apiKey);
        $this->keyId = $parsed['keyId'];
        $this->secret = $parsed['secret'];
        $this->baseUrl = rtrim($options['base_url'] ?? 'https://api.openpay.lk', '/');

        $this->http = new Client([
            'base_uri' => $this->baseUrl,
            'timeout' => $options['timeout'] ?? 30.0,
            'http_errors' => false,
        ]);

        $this->payments = new Payments($this);
        $this->webhooks = new Webhooks($this);
    }

    /**
     * Make an authenticated API request.
     *
     * @return array<string, mixed>
     */
    public function request(string $method, string $path, ?array $body = null): array
    {
        $bodyStr = $body !== null ? json_encode($body, JSON_UNESCAPED_SLASHES) : '';
        $headers = HmacSigner::buildAuthHeaders($this->keyId, $this->secret, $method, $path, $bodyStr);
        $headers['Content-Type'] = 'application/json';

        $options = ['headers' => $headers];
        if ($body !== null) {
            $options['body'] = $bodyStr;
        }

        try {
            $response = $this->http->request($method, $path, $options);
        } catch (RequestException $e) {
            throw new OpenPayException("Request failed: {$e->getMessage()}", 0, $e);
        }

        $statusCode = $response->getStatusCode();
        $data = json_decode((string) $response->getBody(), true) ?? [];

        if ($statusCode >= 400) {
            $error = $data['error'] ?? [];
            $code = $error['code'] ?? 'UNKNOWN_ERROR';
            $message = $error['message'] ?? "HTTP {$statusCode}";

            if ($statusCode === 401) {
                throw new AuthenticationException($message);
            }

            throw new ApiException($code, $message, $statusCode);
        }

        return $data;
    }
}

File: sdks/sdk-java/src/main/java/com/openpay/sdk/OpenPayClient.java
package com.openpay.sdk;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.openpay.sdk.auth.HmacSigner;
import com.openpay.sdk.exceptions.ApiException;
import com.openpay.sdk.exceptions.AuthenticationException;
import com.openpay.sdk.exceptions.OpenPayException;
import com.openpay.sdk.resources.PaymentsResource;
import com.openpay.sdk.resources.WebhooksResource;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;

/**
 * Open Pay API Client.
 *
 * <pre>{@code
 * var client = new OpenPayClient("ak_live_xxx.sk_live_yyy");
 * var payment = client.payments().create(Map.of(
 *     "amount", "1000.00",
 *     "currency", "LKR"
 * ));
 * }</pre>
 */
public class OpenPayClient {

    private static final String DEFAULT_BASE_URL = "https://api.openpay.lk";
    private static final Gson GSON = new Gson();

    private final String keyId;
    private final String secret;
    private final String baseUrl;
    private final HttpClient http;

    private final PaymentsResource payments;
    private final WebhooksResource webhooks;

    public OpenPayClient(String apiKey) {
        this(apiKey, DEFAULT_BASE_URL);
    }

    public OpenPayClient(String apiKey, String baseUrl) {
        String[] parts = HmacSigner.parseApiKey(apiKey);
        this.keyId = parts[0];
        this.secret = parts[1];
        this.baseUrl = baseUrl.replaceAll("/+$", "");
        this.http = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();

        this.payments = new PaymentsResource(this);
        this.webhooks = new WebhooksResource(this);
    }

    public PaymentsResource payments() { return payments; }
    public WebhooksResource webhooks() { return webhooks; }

    /**
     * Make an authenticated API request.
     */
    public JsonObject request(String method, String path, Object body) {
        String bodyStr = body != null ? GSON.toJson(body) : "";
        Map<String, String> authHeaders = HmacSigner.buildAuthHeaders(keyId, secret, method, path, bodyStr);

        HttpRequest.Builder reqBuilder = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + path))
                .timeout(Duration.ofSeconds(30))
                .header("Content-Type", "application/json");

        for (var entry : authHeaders.entrySet()) {
            reqBuilder.header(entry.getKey(), entry.getValue());
        }

        if ("POST".equals(method) || "PUT".equals(method)) {
            reqBuilder.method(method, HttpRequest.BodyPublishers.ofString(bodyStr));
        } else {
            reqBuilder.method(method, HttpRequest.BodyPublishers.noBody());
        }

        try {
            HttpResponse<String> resp = http.send(reqBuilder.build(), HttpResponse.BodyHandlers.ofString());
            JsonObject data = GSON.fromJson(resp.body(), JsonObject.class);

            if (resp.statusCode() >= 400) {
                JsonElement errorElem = data != null ? data.get("error") : null;
                String code = "UNKNOWN_ERROR";
                String message = "HTTP " + resp.statusCode();
                if (errorElem != null && errorElem.isJsonObject()) {
                    JsonObject error = errorElem.getAsJsonObject();
                    if (error.has("code")) code = error.get("code").getAsString();
                    if (error.has("message")) message = error.get("message").getAsString();
                }
                if (resp.statusCode() == 401) {
                    throw new AuthenticationException(message);
                }
                throw new ApiException(code, message, resp.statusCode());
            }

            return data;
        } catch (IOException | InterruptedException e) {
            throw new OpenPayException("Request failed: " + e.getMessage(), e);
        }
    }
}
